File: backend/rails/app/models/evento.rb
# Regras de negócio replicadas de core/models.py (Django) / app/Models/
# Evento.php (Laravel). Assim como em Comunidade, a validação de "forma"
# (tamanhos, formatos, campos cruzados) fica aqui no model. As regras que
# dependem de "quando" a ação acontece (RN-EVT-04 data futura só na
# criação, RN-EVT-07 permissão, RN-EVT-10 bloqueio de evento passado) ficam
# nos controllers — ver app/controllers/eventos_controller.rb e
# app/controllers/api/eventos_controller.rb — porque só se aplicam a
# algumas ações, não a todo `save` (ex.: um evento passado usado como
# fixture de teste não pode falhar validação só por estar no passado).
class Evento < ApplicationRecord
  include UuidPrimaryKey

  PRESENCIAL = "presencial"
  ONLINE = "online"
  HIBRIDO = "hibrido"
  TIPOS = [PRESENCIAL, ONLINE, HIBRIDO].freeze

  belongs_to :comunidade, inverse_of: :eventos
  belongs_to :organizador, class_name: "User"

  before_validation { self.url_online = url_online.presence || "" }

  validates :titulo, presence: true, length: { minimum: 5, maximum: 200 }
  validates :titulo, uniqueness: {
    scope: %i[comunidade_id data], case_sensitive: false,
    message: "Já existe um evento com este título nesta data para esta comunidade."
  }
  validates :descricao, presence: true, length: { minimum: 20 }
  validates :data, presence: true
  validates :hora_inicio, presence: true
  validates :local, presence: true
  validates :tipo, presence: true, inclusion: { in: TIPOS }
  validate :url_online_obrigatorio_para_online_ou_hibrido
  validate :hora_fim_apos_hora_inicio

  private

  def url_online_obrigatorio_para_online_ou_hibrido
    return unless %w[online hibrido].include?(tipo)

    errors.add(:url_online, "é obrigatório quando tipo é 'online' ou 'hibrido'.") if url_online.blank?
  end

  def hora_fim_apos_hora_inicio
    return if hora_fim.blank? || hora_inicio.blank?

    errors.add(:hora_fim, "deve ser posterior a hora_inicio.") if hora_fim <= hora_inicio
  end
end

File: backend/laravel/app/Models/Evento.php
<?php

namespace App\Models;

use App\Support\Auditavel;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Validation\ValidationException;

/**
 * Regras de negócio replicadas de core/models.py (Django) — ver Comunidade
 * para o mesmo padrão (validação centralizada no model via evento saving).
 */
class Evento extends Model
{
    use Auditavel, HasFactory, HasUuids;

    protected $fillable = [
        'titulo', 'descricao', 'data', 'hora_inicio', 'hora_fim', 'local',
        'tipo', 'url_online', 'comunidade_id', 'organizador_id',
    ];

    public const PRESENCIAL = 'presencial';

    public const ONLINE = 'online';

    public const HIBRIDO = 'hibrido';

    public const TIPOS = [self::PRESENCIAL, self::ONLINE, self::HIBRIDO];

    protected $attributes = [
        'url_online' => '',
    ];

    protected function casts(): array
    {
        return [
            'data' => 'date',
        ];
    }

    protected static function booted(): void
    {
        static::saving(function (Evento $evento) {
            // url_online não é anulável no banco (mesmo padrão do blank=True
            // do Django); formulários web enviam '' como null via o
            // middleware ConvertEmptyStringsToNull, então normalizamos aqui.
            $evento->url_online ??= '';

            $evento->validarRegrasDeNegocio();
        });
    }

    public function validarRegrasDeNegocio(): void
    {
        $errors = [];

        if ($this->titulo !== null && ! (mb_strlen($this->titulo) >= 5 && mb_strlen($this->titulo) <= 200)) {
            $errors['titulo'] = 'O título deve ter entre 5 e 200 caracteres.';
        }

        if ($this->descricao !== null && mb_strlen($this->descricao) < 20) {
            $errors['descricao'] = 'A descrição deve ter no mínimo 20 caracteres.';
        }

        if (in_array($this->tipo, [self::ONLINE, self::HIBRIDO], true) && ! $this->url_online) {
            $errors['url_online'] = "url_online é obrigatório quando tipo é 'online' ou 'hibrido'.";
        }

        if ($this->hora_fim && $this->hora_inicio && $this->hora_fim <= $this->hora_inicio) {
            $errors['hora_fim'] = 'hora_fim deve ser posterior a hora_inicio.';
        }

        if ($errors) {
            throw ValidationException::withMessages($errors);
        }
    }

    public function comunidade(): BelongsTo
    {
        return $this->belongsTo(Comunidade::class);
    }

    public function organizador(): BelongsTo
    {
        return $this->belongsTo(User::class, 'organizador_id');
    }
}
